below/tc/src/lib.rs

```diff
@@ -0,0 +1,111 @@
+mod errors;
+mod types;
+
+#[cfg(test)]
+mod test;
+
+use std::collections::BTreeMap;
+
+use netlink_packet_core::{NetlinkHeader, NetlinkMessage, NetlinkPayload, NLM_F_DUMP, NLM_F_REQUEST};
+use netlink_packet_route::RouteNetlinkMessage;
+use netlink_packet_route::tc::TcMessage;
+use netlink_sys::constants::NETLINK_ROUTE;
+use netlink_sys::{Socket, SocketAddr};
+use nix::net::if_;
+
+pub use errors::*;
+pub use types::*;
+
+pub type TcStats = Vec<Tc>;
+pub type Result<T> = std::result::Result<T, TcError>;
+
+/// Get list of all `tc` qdiscs.
+pub fn tc_stats() -> Result<TcStats> {
+    let ifaces = get_interfaces()?;
+    read_tc_stats(ifaces, &get_netlink_qdiscs)
+}
+
+fn read_tc_stats(interfaces: BTreeMap<u32, String>, netlink_retriever: &dyn Fn() -> Result<Vec<TcMessage>>) -> Result<TcStats> {
+    let messages = netlink_retriever()?;
+    let tc_stats = messages
+        .into_iter()
+        .map(|msg| Tc::new(&interfaces, &msg))
+        .collect();
+
+    Ok(tc_stats)
+}
+
+/// Open a netlink socket to retrieve `tc` qdiscs.
+/// The underlying library sends a message of type `RTM_GETQDISC` to the kernel.
+/// The kernel responds with a message of type `RTM_NEWQDISC` for each qdisc.
+fn get_netlink_qdiscs() -> Result<Vec<TcMessage>> {
+    // open a socket
+    let socket = Socket::new(NETLINK_ROUTE).map_err(|e| TcError::Netlink(e.to_string()))?;
+    socket.connect(&SocketAddr::new(0, 0)).map_err(|e| TcError::Netlink(e.to_string()))?;
+
+    // create a netlink request
+    let mut nl_hdr = NetlinkHeader::default();
+    nl_hdr.flags = NLM_F_REQUEST | NLM_F_DUMP;
+    let msg = RouteNetlinkMessage::GetQueueDiscipline(TcMessage::default());
+    let mut packet = NetlinkMessage::new(nl_hdr, NetlinkPayload::from(msg));
+    packet.finalize();
+    let mut buf = vec![0; packet.header.length as usize];
+    packet.serialize(&mut buf[..]);
+
+    // send the request
+    socket.send(&buf[..], 0).map_err(|e| TcError::Netlink(e.to_string()))?;
+
+    // receive the response
+    let mut recv_buf = vec![0; 4096];
+    let mut offset = 0;
+    let mut response = Vec::new();
+    'out: while let Ok(size) = socket.recv(&mut &mut recv_buf[offset..], 0) {
+        loop {
+            let bytes = &recv_buf[offset..];
+            let rx_packet = <NetlinkMessage<RouteNetlinkMessage>>::deserialize(bytes)
+                .map_err(|e| TcError::Netlink(e.to_string()))?;
+            response.push(rx_packet.clone());
+            let payload = rx_packet.payload;
+            if let NetlinkPayload::Error(err) = payload {
+                return Err(TcError::Netlink(err.to_string()));
+            }
+            if let NetlinkPayload::Done(_) = payload {
+                break 'out;
```

Comment:
am i misunderstanding or could this be a `continue`?

Reply:
the outer loop is for each packet received from the kernel
the inner loop is for parsing those packets into netlink object(s)
`Done` implies this is the last packet for the object we requested, so we're done parsing

- from the library [example](https://github.com/rust-netlink/netlink-packet-route/blob/4d5f819b8ca72496fd37b8e885e6d9f77a7545b1/examples/dump_packet_links.rs#L64-L67)